crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4805,6 +4806,45 @@ impl<'db> TypeInferenceBuilder<'db> {
     ) -> Type<'db> {
         let parameters = &*subscript.slice;
         match known_instance {
+            KnownInstanceType::Annotated => {
+                let mut report_invalid_arguments = || {
+                    self.diagnostics.add_lint(
+                        &INVALID_ANNOTATED_ARGUMENTS,
+                        subscript.into(),
+                        format_args!(
+                            "Special form `{}` expected at least 2 arguments (one type and at least one metadata element)",
+                            known_instance.repr(self.db)
+                        ),
+                    );
+                };
+
+                let ast::Expr::Tuple(ast::ExprTuple { elts, .. }) = parameters else {
+                    report_invalid_arguments();
+
+                    // `Annotated[]` with less than two arguments is an error at runtime.
+                    // However, we still treat `Annotated[T]` as `T` here for the purpose of
+                    // giving better diagnostics later on.
+                    // Pyright also does this. Mypy doesn't; it falls back to `Any` instead.
+                    return self.infer_type_expression(parameters);
+                };
+
+                if elts.len() < 2 {
+                    report_invalid_arguments();
+                }
+
+                let [type_expr, metadata @ ..] = &elts[..] else {
+                    self.infer_type_expression(parameters);
+                    return Type::Unknown;
+                };
```

Comment:
```suggestion
                let ast::Expr::Tuple(ast::ExprTuple { elts: arguments, .. }) = parameters else {
                    report_invalid_arguments();

                    // `Annotated[]` with less than two arguments is an error at runtime.
                    // However, we still treat `Annotated[T]` as `T` here for the purpose of
                    // giving better diagnostics later on.
                    // Pyright also does this. Mypy doesn't; it falls back to `Any` instead.
                    return self.infer_type_expression(parameters);
                };

                if arguments.len() < 2 {
                    report_invalid_arguments();
                }

                let [type_expr, metadata @ ..] = &arguments[..] else {
                    self.infer_type_expression(parameters);
                    return Type::Unknown;
                };
```

Ideally we'd also rename the `let parameters = &*subscript.slice;` assignment on line 4807 to something like `let arguments_slice = &*subscript.slice;`. You don't have to do that in this PR, though, if you don't want to.